docs/architecture/adr-069-unordered-account.md

```diff
@@ -0,0 +1,106 @@
+# ADR 069: Un-Ordered Account
+
+## Changelog
+
+* Nov 24, 2023: Initial Draft
+
+## Abstract
+
+Support gaps in account nonce values to support un-ordered(concurrent) transaction inclusion.
+
+## Context
+
+Right now the transactions from a particular sender must be included in strict order because of the nonce value requirement, which makes it tricky to send many transactions concurrently, for example, when a previous pending transaction fails or timeouts, all the following transactions are all blocked. Relayer and exchanges are some typical examples.
+
+The main purpose of nonce value is to protect against replay attack, but for that purpose we only need to make sure the nonce is unique, so we can relax the orderness requirement without lose the uniqueness, in that way we can improve the user experience of concurrent transaction sending.
+
+## Decision
+
+Change the nonce logic to allow gaps to exist, which can be filled by other transactions later, or never filled at all, the prototype implementation use a bitmap to record these gap values.
+
+It's debatable how we should configure the user accounts, for example, should we change the default behavior directly or let user to turn this feature on explicitly, or should we allow user to set different gap capacity for different accounts.
+
+```golang
+const MaxGap = 1024
+
+  type Account struct {
+    ...
+    SequenceNumber int64
++   Gaps *IntSet
+  }
+
+// CheckNonce checks if the input nonce is valid, if yes, modify internal state.
+func(acc *Account) CheckNonce(nonce int64) error {
+  switch {
```

Comment:
I would more godocs to each statement to describe what that case is intended to do. It's not immediately clear to me the last two cases, what is going on.

E.g.

- `nonce >= acct.SequenceNumber + 1:` is for a gap to be filled after the current nonce I think? But it's not clear what the loop is intended to do.
- `nonce < acct.SequenceNumber:` is for a gap that is unfilled, or "in the past". But again, not clear what the logic in the case is intended to do.

Reply:
see if it's better now.